benchexec/resources.py

```diff
@@ -210,118 +218,738 @@ def _get_cpu_cores_per_run0(
+            )
+        else:
+            is_feasible = False
+
+    # calculate if sub_units have to be split to accommodate the runs_per_unit
+    sub_units_per_run = calculate_sub_units_per_run(
+        coreLimit_rounded_up, hierarchy_levels, chosen_level
+    )
+    # number of nodes at subunit-Level / sub_units_per_run
+    if len(hierarchy_levels[chosen_level - 1]) / sub_units_per_run < num_of_threads:
+        if not isTest:
+            sys.exit(
+                f"Cannot split memory regions between runs. "
+                f"Please reduce the number of threads to {math.floor(len(hierarchy_levels[chosen_level-1]) / sub_units_per_run)}."
             )
+        else:
+            is_feasible = False
+
+    return is_feasible
+
+
+def calculate_chosen_level(
+    hierarchy_levels: List[HierarchyLevel], coreLimit_rounded_up: int
+) -> int:
+    """Calculates the hierarchy level necessary so that number of cores at the chosen_level is at least
+    as big as the cores necessary for one thread"""
+    chosen_level = 1
+    # move up in hierarchy as long as the number of cores at the current level is smaller than the coreLimit
+    # if the number of cores at the current level is as big as the coreLimit: exit loop
+    while (
+        chosen_level < len(hierarchy_levels) - 1
+        and len(next(iter(hierarchy_levels[chosen_level].values())))
+        < coreLimit_rounded_up
+    ):
+        chosen_level = chosen_level + 1
+    return chosen_level
+
+
+def calculate_coreLimit_rounded_up(
+    siblings_of_core: HierarchyLevel, coreLimit: int
+) -> int:
+    """coreLimit_rounded_up (int): recalculate # cores for each run accounting for HT"""
+    core_size = len(next(iter(siblings_of_core.values())))
+    # Take value from hierarchy_levels instead from siblings_of_core
+    coreLimit_rounded_up = int(math.ceil(coreLimit / core_size) * core_size)
+    assert coreLimit <= coreLimit_rounded_up < (coreLimit + core_size)
+    return coreLimit_rounded_up
+
+
+def calculate_sub_units_per_run(
+    coreLimit_rounded_up: int,
+    hierarchy_levels: List[HierarchyLevel],
+    chosen_level: int,
+) -> int:
+    """calculate how many sub_units have to be used to accommodate the runs_per_unit"""
+    sub_units_per_run = math.ceil(
+        coreLimit_rounded_up / len(hierarchy_levels[chosen_level - 1][0])
+    )
+    return sub_units_per_run
+
+
+def check_and_add_meta_level(
+    hierarchy_levels: List[HierarchyLevel], allCpus: Dict[int, VirtualCore]
+) -> None:
+    """
+    Adds a meta_level to hierarchy_levels to iterate through all cores (if necessary)
+    """
+    if len(hierarchy_levels[-1]) > 1:
+        top_level_cores = []
+        for node in hierarchy_levels[-1]:
+            top_level_cores.extend(hierarchy_levels[-1][node])
+        hierarchy_levels.append({0: top_level_cores})
+        for cpu_nr in allCpus:
+            allCpus[cpu_nr].memory_regions.append(0)
```

Comment:
It seems this adds the memory region `0` to every core. But isn't this blatantly wrong? Why is this here, without any explanation?

Reply:
Yes, it adds a `0` to the memory region list of each core, because on the highest hierarchy level that we just added, all cores belong to the unit with id `0` (see linie 413). will add an explanation in the next commit.

Reply:
Resolveed by a00ede6cd284523173d5f9debbe1f900cb2a77be.